### ISSUE TITLE
Can't train a model having bias addition layer Add()

### ISSUE THREAD

**Issue by kumarutkarsh1248** [mlpack/mlpack#3666]
instead of using Linear() layer I am trying to train the model with LinearNoBIas() and Add() layer,
but getting error while doing so 

my full code:
```

#include "mlpack.hpp"
using namespace std;
using namespace mlpack;


int main()
{
        arma::mat matrix(100, 2);
        for (int i = 0; i < 100; i++) 
        {
                matrix(i, 0) = i;
                matrix(i, 1) = 0.7*i + 3;
        }
        matrix = matrix.t();

        // spliting the data into training and validation dataset
        arma::mat train, valid;
        float RATIO = 0.3;
        mlpack::data::Split(matrix, train, valid, RATIO);

        arma::mat Xtrain, Xvalid, ytrain, yvalid;
        Xtrain = train.row(0);
        ytrain = train.row(1);

        // Xvalid and yvalid assignment corrected
        Xvalid = valid.row(0);
        yvalid = valid.row(1);

        mlpack::FFN<mlpack::MeanSquaredError, mlpack::RandomInitialization> model;

        model.Add(new LinearNoBias(1));
        model.Add(new Add());
        // model.Add(new Linear(3));
        // model.Add(new Linear(4));

        vector<size_t> v= {1};
        model.InputDimensions() = v;   // it basically takes the input features of one data point
        model.Reset();

        for (int i = 0; i < 1000; i++)
        {
                ens::AdaDelta optimizer(0.09, 2, 0.99, 1e-8, Xtrain.n_cols, 1e-9, true);
                model.Train
                (
                        Xtrain,
                        ytrain,
                        optimizer,
                        ens::PrintLoss(),
                        ens::ProgressBar(),
                        ens::EarlyStopAtMinLoss(20)
                );
        }
        for(auto layer : model.Network())
        {
                layer->Parameters().print("layer");
        }
        return 0;
}
```

i am getting error:
terminate called after throwing an instance of 'std::logic_error'
  what():  Mat::init(): mismatch between size of auxiliary memory and requested size
make: *** [makefile:21: run] Aborted (core dumped)

while debugging, i found that the program throw exception at

```
//! print a message to get_cerr_stream() and throw logic_error exception
template<typename T1>
arma_cold
arma_noinline
static
void
arma_stop_logic_error(const T1& x)
  {
  #if (defined(ARMA_PRINT_EXCEPTIONS) && defined(ARMA_PRINT_ERRORS))
    {
    get_cerr_stream() << "\nerror: " << x << std::endl;
    }
  #endif
  
  throw std::logic_error( std::string(x) );
  }
```

its in the armadillo debug.hpp file

**Comment by rcurtin** [comment]
That's where Armadillo throws the exception, yeah, but the actual error will be a couple of stack frames higher inside the mlpack codebase.

**Comment by kumarutkarsh1248** [comment]
okay so the error is occuring here

```
template<typename MatType>
void AddType<MatType>::Gradient(
    const MatType& /* input */,
    const MatType& error,
    MatType& gradient)
{
  gradient = error;
}
```

one thing that i have noticed is that error has mem_state: 0 while gradient has mem_state: 2

![Screenshot from 2024-03-27 08-44-51](https://github.com/mlpack/mlpack/assets/118059940/35d8fcc2-49f4-4ca9-a412-d476003ff8c6)


can it be the cause of error??

**Comment by rcurtin** [comment]
The `mem_state` indicates whether or not a matrix is an alias; so, here, `gradient` is an alias, but `error` is not.  I noticed that `gradient` is of size `1x1` though, but I think this is incorrect: aren't your input features 100-dimensional?  I think your call to `InputDimensions()` is incorrect.  (I also think it's unnecessary; by default the `FFN` class will assume that its given data consists of one-dimensional points whose dimensionality is equal to the number of rows in the given data matrix.)

I think there may be something wrong with `LinearNoBias` but I'm not 100% sure.  If you want to debug it, I would suggest changing `LinearNoBias(1)` to `Linear(1)`, seeing what is different in the sizes of matrices in the `Linear` class when `Gradient()` is called, and go from there. :+1:

**Comment by MarkFischinger** [comment]
I've been working on resolving this issue too, but I haven't had quite enough time these past few days to completely figure it out.
From what I can tell, the problem appears to be rooted solely in the `AddLayer`. Interestingly, replacing a Linear(1) for a LinearNoBias(1) isn’t effective either, as it also leads to the same error in my code, which leads me to the conclusion that the LinearNoBias function should be working correctly:

```cpp
#include "mlpack.hpp"
using namespace std;
using namespace mlpack;


int main() {
	arma::arma_rng::set_seed(1000);
	arma::mat trainData = arma::randu(2, 100);
	arma::mat trainLabels = arma::randu(1, 100);

	arma::cout << "n rows="<<trainData.n_rows <<" n cols="<< trainData.n_cols << arma::endl;
	arma::cout << "n rows="<< trainLabels.n_rows <<" n cols="<< trainLabels.n_cols << arma::endl;


	arma::uword i,j;
	std::cout<<"First 10 Labels:"<<std::endl;
	for(i=0;i<10;i++){
	arma::cout << trainLabels(0,i) << arma::endl;
	}

	j=0;
	std::cout<<"Features for observation: "<<j<<std::endl;
	for(i=0;i<trainData.n_rows;i++){
	arma::cout << trainData(i,j) << arma::endl;
	}
	j=1;
	std::cout<<"Features for observation:"<<j<<std::endl;
	for(i=0;i<trainData.n_rows;i++){
	arma::cout << trainData(i,1) << arma::endl;
	}

	mlpack::FFN<mlpack::MeanSquaredError, mlpack::RandomInitialization> model;

	const size_t inputSize=trainData.n_rows; // 9
	const size_t outputSize=1;
	const size_t hiddenLayerSize=2;

	model.Add(new Linear(inputSize));
	model.Add(new Linear(hiddenLayerSize));
	model.Add(new Linear(outputSize));
	model.Add(new Add());

	model.InputDimensions() = {inputSize};
	model.Reset();

    	ens::AdaDelta optimizer(0.09, 2, 0.99, 1e-8, trainData.n_cols, 1e-9, true);
    	model.Train
    	(
            	trainData,
            	trainLabels,
            	optimizer,
            	ens::PrintLoss(),
            	ens::ProgressBar(),
            	ens::EarlyStopAtMinLoss(20)
    	);
    	for(auto layer : model.Network())
    	{
            	layer->Parameters().print("layer");
    	}
    	return 0;

}
```
```
Error size: 1x2
Gradient size: 1x1
terminate called after throwing an instance of 'std::logic_error'
  what():  Mat::init(): mismatch between size of auxiliary memory and requested size
Aborted
```

He should probably set the InputDimensions(), as not doing so might lead to this error: `terminate called after throwing an instance of 'std::invalid_argument' what(): FFN::Reset(): cannot reset network when no input dimensionality is given, and InputDimensions() has not been set! Aborted.`

Resizing the `gradient` inside the `AddType<MatType>::Gradient` isn't an option, as its memory allocation is initialized by the FNN and MultiLayers above. I'm not sure just yet when I'll be able to fix this as my schedule is pretty packed in the upcoming weeks.

**Comment by kumarutkarsh1248** [comment]
@MarkFischinger 
I completely agree with you. I have tried various layer combinations and come to the conclusion that the problem specifically lies with AddLayer(). Additionally, the error occurs while executing the line 
`gradient = error`
I believe that in Armadillo, assignment between matrices of different sizes is allowed, so resizing gradient will not solve the issue.

**Comment by kumarutkarsh1248** [comment]
that error is generated from this line:
  `  arma_debug_check( (t_mem_state == 2), "Mat::init(): mismatch between size of auxiliary memory and requested size" );
  `

**Comment by rcurtin** [comment]
I spent some time with the code today.  My comment about `InputDimensions()` was not relevant because I missed this line when reading the test program:

```
        Xtrain = train.row(0);
```

So, the input is indeed just one feature.

As an aside, as I mentioned, you do not need to specify `InputDimensions()` if the input is meant to consist of vector-type points (e.g., an image input is 2-dimensional and you would need to specify it manually).  If you want to call `Reset()` *without* setting `InputDimensions()`, you have to call `Reset(inputDimensionality)`.  I recognize that there is not good Markdown documentation for the `FFN` class yet, but you can see this directly in the comments for the `Reset()` method:

 * https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/ffn.hpp#L220
 * https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/ffn.hpp#L248

Back to the actual issue at hand:

The backtrace given with gdb points to the function `Add::Gradient()`.  From that point it becomes fairly clear that the implementation is not accounting for a batch size greater than 1.  I opened a fix PR in #3679; there is more of an explanation there.

Anyway, thank you for opening the issue, and I'm glad we could track the bug down. :+1:

**Comment by rcurtin** [comment]
Fixed in #3679. :+1:
